src/register-service-worker.js

```diff
@@ -0,0 +1,162 @@
+      }));
+    });
+  }
+
+  /**
+   * Compares the copy of a manifest obtained from fetch() with the copy stored
+   * in IndexedDB. If they differ, it kicks off the manifest update process.
+   *
+   * @param {DB} db
+   * @param {String} manifestUrl
+   * @returns {Promise.<T>}
+   */
+  function handlePossibleManifestUpdate(db, manifestUrl) {
+    var tx = db.transaction(constants.OBJECT_STORES.MANIFEST_URL_TO_CONTENTS);
+    var store = tx.objectStore(
+      constants.OBJECT_STORES.MANIFEST_URL_TO_CONTENTS);
+
+    return Promise.all([
+      fetch(manifestUrl).then(function(manifestResponse) {
+        return manifestResponse.text();
+      }),
+      store.get(manifestUrl).then(function(idbEntryForManifest) {
+        return idbEntryForManifest ? idbEntryForManifest.text : '';
+      })
+    ]).then(function(manifestTexts) {
+      // manifestTexts[0] is the text content from the fetch().
+      // manifestTexts[1] is the text content from IDB.
+      if (manifestTexts[0] !== manifestTexts[1]) {
+        return performManifestUpdate(db, manifestUrl, manifestTexts[0]);
+      }
+    });
+  }
+
+  /**
+   * Parses the newest manifest text into the format described at
+   * https://www.npmjs.com/package/parse-appcache-manifest
+   * The parsed manifest is stored in IndexedDB.
+   * This also calls addToCache() to cache the relevant URLs from the manifest.
+   *
+   * @param {DB} db
+   * @param {String} manifestUrl
+   * @param {String} manifestText
+   * @returns {Promise.<T>}
+   */
+  function performManifestUpdate(db, manifestUrl, manifestText) {
+    var parseAppCacheManifest = require('parse-appcache-manifest');
+    var parsedManifest = makeManifestUrlsAbsolute(manifestUrl,
+      parseAppCacheManifest(manifestText));
+
+    var tx = db.transaction(constants.OBJECT_STORES.MANIFEST_URL_TO_CONTENTS,
+      'readwrite');
+    var store = tx.objectStore(
+      constants.OBJECT_STORES.MANIFEST_URL_TO_CONTENTS);
+
+    var fallbackUrls = Object.keys(parsedManifest.fallback).map(function(key) {
+      return parsedManifest.fallback[key];
+    });
+
+    return Promise.all([
+      store.put({
+        text: manifestText,
+        parsed: parsedManifest
+      }, manifestUrl),
+      addToCache(parsedManifest.cache.concat(fallbackUrls))
+    ]);
+  }
+
+  /**
+   * Updates IndexedDB to indicate that the current page's URL is associated
+   * with the AppCache manifest at manifestUrl.
+   * It also adds the current page to the cache, matching the implicit
+   * cache-as-you-go behavior you get with AppCache.
+   *
+   * @param {DB} db
+   * @param {String} manifestUrl
+   * @returns {Promise.<T>}
+   */
+  function updateManifestAssociationForCurrentPage(db, manifestUrl) {
+    var tx = db.transaction(constants.OBJECT_STORES.PATH_TO_MANIFEST,
+      'readwrite');
+    var store = tx.objectStore(constants.OBJECT_STORES.PATH_TO_MANIFEST);
+
+    return Promise.all([
+      store.put(manifestUrl, location.href),
```

Comment:
Annoyingly this can resolve, but the transaction can still fail. `tx.complete` is a promise for the IDB transaction.

I don't know why IDB makes a distinction between the two, I'm sure there's some complicated reason.